AdminServer/appscale/admin/__init__.py

```diff
@@ -0,0 +1,404 @@
+    except UAException:
+      message = 'Unable to ensure project exists: {}'.format(project_id)
+      logging.exception(message)
+      raise CustomHTTPError(HTTPCodes.INTERNAL_ERROR, message=message)
+
+  def ensure_user_is_owner(self, project_id, user):
+    """ Ensures a user is the owner of a project.
+    
+    Args:
+      project_id: A string specifying a project ID.
+      user: A string specifying a user's email address.
+    Raises:
+      CustomHTTPError if the user is not the owner.
+    """
+    try:
+      project_metadata = self.ua_client.get_app_data(project_id)
+    except UAException:
+      message = 'Unable to retrieve project metadata'
+      logging.exception(message)
+      raise CustomHTTPError(HTTPCodes.INTERNAL_ERROR, message=message)
+
+    if 'owner' not in project_metadata:
+      raise CustomHTTPError(HTTPCodes.INTERNAL_ERROR,
+                            message='Project owner not defined')
+
+    if project_metadata['owner'] != user:
+      message = 'User is not project owner: {}'.format(user)
+      raise CustomHTTPError(HTTPCodes.FORBIDDEN, message=message)
+
+  def begin_deploy(self, project_id, source_path):
+    """ Triggers the deployment process.
+    
+    Args:
+      project_id: A string specifying a project ID.
+      source_path: A string specifying the location of the version's source.
+    Raises:
+      CustomHTTPError if unable to start the deployment process.
+    """
+    try:
+      self.acc.done_uploading(project_id, source_path)
+    except AppControllerException as error:
+      message = 'Error while setting sourceUrl: {}'.format(error)
+      raise CustomHTTPError(HTTPCodes.INTERNAL_ERROR, message=message)
+
+    try:
+      self.acc.update([project_id])
+    except AppControllerException as error:
+      message = 'Error while updating application: {}'.format(error)
+      raise CustomHTTPError(HTTPCodes.INTERNAL_ERROR, message=message)
+
+  def post(self, project_id, service_id):
+    """ Creates or updates a version.
+    
+    Args:
+      project_id: A string specifying a project ID.
+      service_id: A string specifying a service ID.
+    """
+    self.authenticate()
+    user = self.get_current_user()
+    version = self.version_from_payload()
+    project_exists = self.project_exists(project_id)
+    if not project_exists:
+      self.create_project(project_id, user, version['runtime'])
+
+    if service_id != constants.DEFAULT_SERVICE:
+      raise CustomHTTPError(HTTPCodes.BAD_REQUEST, message='Invalid service')
+
+    self.ensure_user_is_owner(project_id, user)
+
+    # Strip protocol prefix from sourceUrl.
+    proto_prefix = 'file://'
+    source_path = version['deployment']['zip']['sourceUrl']
+    if source_path.startswith(proto_prefix):
+      source_path = source_path[len(proto_prefix):]
+
+    self.begin_deploy(project_id, source_path)
+
+    operation_id = str(uuid.uuid4())
+    insert_time = datetime.datetime.utcnow()
+
+    operations[operation_id] = {
```

Comment:
operation entity has strict structure and it's used many times. Doesn't it worth to create a model for this? It would reduce amount of representative string constants..